hail/src/main/scala/is/hail/expr/ir/ExtractIntervalFilters.scala

```diff
@@ -2,32 +2,75 @@ package is.hail.expr.ir
 
 import is.hail.backend.ExecuteContext
 import is.hail.annotations.IntervalEndpointOrdering
+import is.hail.rvd.PartitionBoundOrdering
 import is.hail.types.virtual._
 import is.hail.utils.{FastSeq, Interval, IntervalEndpoint, _}
 import is.hail.variant.{Locus, ReferenceGenome}
 import org.apache.spark.sql.Row
 
+import scala.Option.option2Iterable
+
 object ExtractIntervalFilters {
 
   val MAX_LITERAL_SIZE = 4096
 
-  case class ExtractionState(ctx: ExecuteContext, rowRef: Ref, keyFields: IndexedSeq[String]) {
+  object ExtractionState {
+    def apply(
+      ctx: ExecuteContext,
+      rowRef: Ref,
+      keyFields: IndexedSeq[String]
+    ): ExtractionState = {
+      new ExtractionState(ctx, Env.empty[IR], rowRef, keyFields)
+    }
+  }
+
+  class ExtractionState(
+    val ctx: ExecuteContext,
+    env: Env[IR],
+    rowRef: Ref,
+    keyFields: IndexedSeq[String]
+  ) {
     val rowType: TStruct = rowRef.typ.asInstanceOf[TStruct]
     val rowKeyType: TStruct = rowType.select(keyFields)._1
     val firstKeyType: Type = rowKeyType.types.head
-    val iOrd: IntervalEndpointOrdering = rowKeyType.ordering(ctx.stateManager).intervalEndpointOrdering
+    val iOrd: IntervalEndpointOrdering = PartitionBoundOrdering(ctx, rowKeyType).intervalEndpointOrdering
 
     def getReferenceGenome(rg: String): ReferenceGenome = ctx.stateManager.referenceGenomes(rg)
 
-    def isFirstKey(ir: IR): Boolean = ir == GetField(rowRef, rowKeyType.fieldNames.head)
+    def bind(name: String, v: IR): ExtractionState =
+      new ExtractionState(ctx, env.bind(name, v), rowRef, keyFields)
+
+    private def isRowRef(ir: IR): Boolean = ir match {
+      case ref: Ref =>
+        ref == rowRef || env.lookupOption(ref.name).exists(isRowRef)
```

Comment:
Something has puzzled me for a while and that's refs of refs.
Are they a thing (this code seems like it)?
Should they be a thing?
What do they mean?